### PR TITLE
libkbfs: (temporarily) track directory write operations in an in-memory cache

### PR DESCRIPTION
**Pull request by strib**
This PR is laying the groundwork for buffering directory write operations in-memory and batching them together in a single journal write.  For the moment, all it does it save them in memory *for the duration of the write operation*, applying them whenever the block is accessed in the meantime, and then clears the cache.  But it gets much of the function call and data structure work in place for making the batching possible in a future PR.

Note that in a future commit I'm planning to get rid of `folderBlockOps.deCache` completely, in favor of just using `DirtyBlockCache`.  I can't do that yet, because if there are multiple dirty files being written to at once in the same directory, they will be Sync'd in different calls, and thus we don't want the dirty dir block containing both dirty entries flushed when only one of them is Sync'd.  In the future, I'm going to remove the ability to Sync a single dirty file at a time (at least for the near term), and so we'll always be syncing everything at once (in order to maintain our current ordering guarantees, which is actually stricter than POSIX requires).

Issue: KBFS-2071

### CODE REVIEW

File: libkbfs/folder_branch_ops.go [hunk 1]
```diff
@@ -2494,16 +2500,80 @@ func (fbo *folderBranchOps) createEntryLocked(
...
+		true, true, zeroPtr, excl)
+	if excl == WithExcl && errors.Cause(err) == errNoUpdatesWhileDirty {
+		// If an exclusive write hits a conflict, it will try to
+		// update, but won't be able to because of the dirty directory
+		// entries.  We need to clean up the dirty entries here first
+		// before trying to apply the updates again.
+		fbo.log.CDebugf(ctx, "Clearing dirty entries before applying new "+
+			"updates for exclusive write")
+		cleanupFn()
+		err = fbo.getAndApplyMDUpdates(ctx, lState, fbo.applyMDUpdatesLocked)
```
Comment (strib) [inline libkbfs/folder_branch_ops.go]: This part is pretty ugly but I'm not sure how else to clean out the dirty data to allow the updates to be applied.

**Comment by strib** [comment]
(By the way, I'll probably merge once I get 2 reviews -- 3 seems like overkill for blocking the PR.  But the more feedback, the better of course!)

**Comment by jzila** [comment]
Nice! FWIW I agree with your `Sync` ordering guarantee: IMO it's a logical outcome from content addressable hierarchy: a sync for any internal tree node should encompass any pending changes for descendant nodes.

**Comment by songgao** [comment]
> I'm going to remove the ability to Sync a single dirty file at a time (at least for the near term), and so we'll always be syncing everything at once

Removing individual syncs sounds nice! Would remove some complexity. Two minor questions: 1) Does this alter the behavior of syncing a file that has been removed, using a handle created before the removal? 2) I guess "sync everything" means sync everything within the same folder branch. Is there any concerns / would it cause confusions for user?

**Comment by strib** [comment]
> 1) Does this alter the behavior of syncing a file that has been removed, using a handle created before the removal?

Not really, except that an explicit fsync on a file handle that has been removed will trigger a sync of everything in the folder-branch (still ignoring the removed file).  [Not true in this PR, I'm talking about once I've finished all this work.]

> 2) I guess "sync everything" means sync everything within the same folder branch. Is there any concerns / would it cause confusions for user?

Correct, everything in the folder branch.  Right now there isn't a way in POSIX for the user to say "sync the entire filesystem" (as far as I know), and FUSE doesn't support such a thing, so the application is always doing an fsync of a particular node.  So in that sense, it shouldn't cause any confusion since the app can really only sync within a folder-branch anyway.  Not sure yet if we should support `fsync(/keybase)` or `fsync(/keybase/private)` for syncing all of the dirty folder-branches.  I probably won't at first, and will continue to let FUSE think it is unimplemented for those nodes.

File: libkbfs/folder_block_ops.go [hunk 2]
```diff
@@ -803,13 +816,167 @@ func (fbo *folderBlockOps) GetDir(
...
+	case exAttr:
+		fileEntry.dirEntry.Type = realEntry.Type
+	case mtimeAttr:
+		fileEntry.dirEntry.Mtime = realEntry.Mtime
+	}
+	fileEntry.dirEntry.Ctime = realEntry.Ctime
+	fbo.deCache[ref] = fileEntry
+}
+
+// SetAttrInDirEntryInCache removes an entry fron the given directory
```
Comment (songgao) [inline libkbfs/folder_block_ops.go]: s/fron/from

File: libkbfs/folder_branch_ops.go [hunk 3]
```diff
@@ -2494,16 +2500,80 @@ func (fbo *folderBranchOps) createEntryLocked(
...
+		ctx, lState, md, newBlock, dirPath, name, entryType,
+		true, true, zeroPtr, excl)
+	if excl == WithExcl && errors.Cause(err) == errNoUpdatesWhileDirty {
+		// If an exclusive write hits a conflict, it will try to
+		// update, but won't be able to because of the dirty directory
+		// entries.  We need to clean up the dirty entries here first
+		// before trying to apply the updates again.
+		fbo.log.CDebugf(ctx, "Clearing dirty entries before applying new "+
+			"updates for exclusive write")
+		cleanupFn()
```
Comment (strib) [inline libkbfs/folder_branch_ops.go]: Yeah it's a little confusing.  But below, by returning `ExclOnUnmergedError{}`, we trigger a retry of the whole operation in  `doMDWriteWithRetry` (the same way that it's usually triggered by `finalizeMDWriteLocked`).  I can update the comment above to make that more clear.

Comment (songgao) [inline libkbfs/folder_branch_ops.go]: Ah I see; Thanks! Could the dirty entries include anything from previous FS calls though? As I understand this throws away all unflushed stuff; doesn't it mean we'd lose any unflushed changes on a failed exclusive create?

Comment (strib) [inline libkbfs/folder_branch_ops.go]: Good question, but no, not in this PR.  The reason is that this operation holds `mdWriterLock`, as does every operation that calls these functions.  And all of those functions clear out their own changes before they release the lock.  So we know for sure that the only outstanding dirty changes are the ones from this operation.

It's true that in a future PR we'll need a better way of undoing a single set of changes -- not just for the excl write case, but also any retry case.  I'll be thinking about how to do that better, but I don't think it needs to hold up this PR.

Comment (songgao) [inline libkbfs/folder_branch_ops.go]: Ah that totally makes sense. Thanks!

File: libkbfs/folder_branch_ops.go [hunk 4]
```diff
@@ -2494,16 +2500,82 @@ func (fbo *folderBranchOps) createEntryLocked(
...
+			Type: entryType,
+			Size: 0,
+		},
+	}
+	fbo.blocks.AddDirEntryInCache(lState, dirPath, name, de)
+	fbo.dirOps = append(fbo.dirOps, cachedDirOp{co, []Node{dir, node}})
+
+	cleanupFn := func() {
+		// Until KBFS-2076 is done, clear out the cached dir data manually.
+		fbo.dirOps = nil
```
Comment (jzila) [inline libkbfs/folder_branch_ops.go]: So for now, you're wiping `dirOps` after every operation. I assume that cleanup will go away, and the batching will take care of that eventually?

File: libkbfs/folder_branch_ops.go [hunk 5]
```diff
@@ -2494,16 +2500,82 @@ func (fbo *folderBranchOps) createEntryLocked(
...
+		err = fbo.getAndApplyMDUpdates(ctx, lState, fbo.applyMDUpdatesLocked)
+		if err != nil {
+			return nil, DirEntry{}, err
+		}
+		return nil, DirEntry{}, ExclOnUnmergedError{}
+	} else if err != nil {
+		return nil, DirEntry{}, err
+	}
+
+	{
```
Comment (jzila) [inline libkbfs/folder_branch_ops.go]: I didn't realize Go lets you have naked blocks like this. Is that necessary just to keep in mind for the future that this'll be removed?

File: libkbfs/folder_block_ops.go [hunk 6]
```diff
@@ -1417,7 +1624,9 @@ func (fbo *folderBlockOps) writeDataLocked(
 	// the `deCache` is used to determine whether there are any dirty
 	// files.  TODO: combine `deCache` with `dirtyFiles` and
 	// `unrefCache`.
-	fbo.deCache[file.tailPointer().Ref()] = newDe
+	cacheEntry := fbo.deCache[file.tailPointer().Ref()]
+	cacheEntry.dirEntry = newDe
```
Comment (strib) [inline libkbfs/folder_block_ops.go]: If it wasn't in the cache before, then cacheEntry (which is a non-pointer type) is the struct zero value, with all the fields at zero values.  So this new code should be exactly equivalent to the old code.

File: libkbfs/folder_branch_ops.go (hunk 4 above)
Comment (strib) [inline libkbfs/folder_branch_ops.go]: Yep, exactly, that's the plan.

File: libkbfs/folder_branch_ops.go (hunk 5 above)
Comment (strib) [inline libkbfs/folder_branch_ops.go]: It's not _necessary_ per se, I was just using it to help me see what needed cleaning up later, and to make sure no variables declared in this to-be-removed code are used elsewhere in the function.  If it bugs you or you want more commenting around it, let me know.

Comment (jzila) [inline libkbfs/folder_branch_ops.go]: Naw, it's fine. Just making sure that was the reason.

File: libkbfs/folder_block_ops.go [hunk 7]
```diff
@@ -123,6 +123,19 @@ func (si *syncInfo) mergeUnrefCache(md *RootMetadata) {
...
 
+type deCacheEntry struct {
+	// dirEntry is the dirty directory entry corresponding to the
+	// BlockPointer that maps to this struct.
+	dirEntry DirEntry
+	// adds is a map of the pointers for new entry names that have
+	// been added to the DirBlock for the BlockPointer that maps to
+	// this struct.
+	adds map[string]BlockPointer
+	// dels is a set of the name that have been removed from the
```
Comment (akalin-keybase) [inline libkbfs/folder_block_ops.go]: name -> names

File: libkbfs/folder_block_ops.go [hunk 8]
```diff
@@ -803,39 +816,237 @@ func (fbo *folderBlockOps) GetDir(
...
+	// Ctime on a rename.
+	cacheEntry, ok := fbo.deCache[newDe.Ref()]
+	if ok && cacheEntry.dirEntry.IsInitialized() {
+		cacheEntry.dirEntry.Ctime = newDe.Ctime
+	} else {
+		cacheEntry.dirEntry = newDe
+		fbo.deCache[newDe.Ref()] = cacheEntry
+		deleteTargetDirEntry = true
+	}
+	fbo.deCache[newDe.Ref()] = cacheEntry
```
Comment (akalin-keybase) [inline libkbfs/folder_block_ops.go]: is this on purpose? seems to be the same as in line 898, so one of the lines must be redundant...

File: libkbfs/folder_branch_ops.go [hunk 9]
```diff
@@ -3967,6 +4101,8 @@ func (fbo *folderBranchOps) getCurrMDRevision(
 
 type applyMDUpdatesFunc func(context.Context, *lockState, []ImmutableRootMetadata) error
 
+var errNoUpdatesWhileDirty = errors.New("Ignoring MD updates while writes are dirty")
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: Nit: can you use a new type instead? The problem with `errors.New` is that it records the stack at instantiation time, which might be confusing (probably still printed even if we call WithStack below)

File: libkbfs/folder_block_ops.go (hunk 8 above)
Comment (strib) [inline libkbfs/folder_block_ops.go]: Good catch. Getting rid of the one above.

**Pull request merged**